### ISSUE TITLE
provide a way to predownload binary

### ISSUE THREAD

**Issue by gengjiawen** [prisma/prisma#19994]
## Problem

When I execute `npx prisma migrate deploy`, url is blocked by server firewall (npm is in white list)

```console
 Error: request to https://binaries.prisma.sh/all_commits/b20ead4d3ab9e78ac112966e242ded703f4a052c/debian-openssl-1.1.x/migration-engine.gz.sha256 failed, reason: read ECONNRESET
```

## Suggested solution

Can those binary put it in npm lib or integrate with prisma itself, not always download from the clound

**Comment by Jolg42** [comment]
Hi @gengjiawen I think you are using `4.16` version here, so you could try this: https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_engines_checksum_ignore_missing

Of course this won't work if the engine was not downloaded before but I suspect that the engines were downloaded but you're hitting an issue with how the checksum files are handled.

Let us know if this works 👀 

If it does not work, I'm curious to know more about how this app is built and deployed.

**Comment by gengjiawen** [comment]
```console 
Error: request to https://binaries.prisma.sh/all_commits/b20ead4d3ab9e78ac112966e242ded703f4a052c/debian-openssl-1.1.x/libquery_engine.so.node.gz failed, reason: read ECONNRESET
```

**Comment by Jolg42** [comment]
@gengjiawen So how & in which environment do you build and deploy your app?

**Comment by gengjiawen** [comment]
> @gengjiawen So how & in which environment do you build and deploy your app?

Our server firewall only allow npm and docker. So can those binaries put into npm default bundled.

**Comment by Jolg42** [comment]
How & in which environment do you build and deploy your app?
If you're using Docker, it would help if you could share your Dockerfile.

The engines are normally downloaded during a `postinstall` hook by the package manager, and nothing needs to be done to "pre-dowload" them.
Did you disable scripts, maybe? (with `npm install --ignore-scripts`)

If you're using Docker I think you'll need to copy over the `node_modules` directory.
The engines are downloaded to `./node_modules/@prisma/engines`

Like https://notiz.dev/blog/dockerizing-nestjs-with-prisma-and-postgresql
```
COPY --from=builder /app/node_modules ./node_modules
```

Alternatively, you could allow `https://binaries.prisma.sh` in your firewall settings, this is the default location for Prisma engines, see https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_engines_mirror

**Comment by Jolg42** [comment]
Note that most CLI commands check and download engines if needed.
For example, running `prisma -v` will check and download engines if necessary and output the versions.

**Comment by janpio** [comment]
> Can those binary put it in npm lib or integrate with prisma itself, not always download from the clound

No, as these files are different depending on your operating system and some configuration, and that would mean that the Npm package needs to include all of these and would be massive.

If you can not download the files via our CLI and tooling, you will have to do so manually and put them into a known location and then configure that custom location for the files: https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#custom-engine-file-locations

**Comment by nedgar** [comment]
There's also a gap in the certificate chain (another side effect of our corporate firewall).
e.g. in the `hello-prisma` example directory, `npx prisma -v` fails with `request to ... failed, reason: unable to verify the first certificate`:
```bash
% DEBUG="prisma:*" npx prisma -v    
  prisma:engines  binaries to download libquery-engine, migration-engine +0ms
  prisma:download  file /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/libquery_engine-darwin.dylib.node does not exist and must be downloaded +0ms
  prisma:download  file /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/migration-engine-darwin does not exist and must be downloaded +0ms
  prisma:download  Downloading https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.gz to /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/libquery_engine-darwin.dylib.node +2ms
> Downloading Prisma engines for Node-API for darwin [                    ] 0%
  prisma:download  Downloading https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/migration-engine.gz to /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/migration-engine-darwin +11ms
Error: request to https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.sha256 failed, reason: unable to verify the first certificate
```

With `export NODE_TLS_REJECT_UNAUTHORIZED=0`, it gets further, but (as mentioned above) .gz files are still blocked.

```bash
% DEBUG="prisma:*" npx prisma -v       
  prisma:engines  binaries to download libquery-engine, migration-engine +0ms
  prisma:download  file /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/libquery_engine-darwin.dylib.node does not exist and must be downloaded +0ms
  prisma:download  file /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/migration-engine-darwin does not exist and must be downloaded +1ms
  prisma:download  Downloading https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/migration-engine.gz to /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/migration-engine-darwin +2ms
> Downloading Prisma engines for Node-API for darwin [                    ] 0%  prisma:download  Downloading https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.gz to /Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/libquery_engine-darwin.dylib.node +13ms
  prisma:downloadZip  Error: Forbidden https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.gz
    at result.retries (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:22930:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RetryOperation._fn (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:20131:19) {
  attemptNumber: 1,
  retriesLeft: 2
} +0ms
  prisma:downloadZip  Error: Forbidden https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/migration-engine.gz
    at result.retries (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:22930:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RetryOperation._fn (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:20131:19) {
  attemptNumber: 1,
  retriesLeft: 2
} +4ms
  prisma:downloadZip  Error: Forbidden https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.gz
    at result.retries (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:22930:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RetryOperation._fn (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:20131:19) {
  attemptNumber: 2,
  retriesLeft: 1
} +1s
  prisma:downloadZip  Error: Forbidden https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/migration-engine.gz
    at result.retries (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:22930:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RetryOperation._fn (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:20131:19) {
  attemptNumber: 2,
  retriesLeft: 1
} +2ms
  prisma:downloadZip  Error: Forbidden https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.gz
    at result.retries (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:22930:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RetryOperation._fn (/Users/nick.alexander.edgar/src/github.com/nedgar/hello-prisma/node_modules/@prisma/engines/dist/index.js:20131:19) {
  attemptNumber: 3,
  retriesLeft: 0
} +3s
Error: Forbidden https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/darwin/libquery_engine.dylib.node.gz
```

**Comment by ludralph** [comment]
Closing this in favour of [12593](https://github.com/prisma/prisma/issues/12593)
